src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSSynchronizationContext.cs

```diff
@@ -47,42 +41,30 @@ public WorkItem(SendOrPostCallback callback, object? data, ManualResetEventSlim?
             }
         }
 
-        internal JSSynchronizationContext(Thread targetThread, IntPtr targetThreadId)
-            : this(
-                targetThread, targetThreadId,
-                Channel.CreateUnbounded<WorkItem>(
-                    new UnboundedChannelOptions { SingleWriter = false, SingleReader = true, AllowSynchronousContinuations = true }
-                )
-            )
-        {
-        }
-
-        internal static void AssertWebWorkerContext()
-        {
-#if FEATURE_WASM_THREADS
-            if (CurrentJSSynchronizationContext == null)
-            {
-                throw new InvalidOperationException("Please use dedicated worker for working with JavaScript interop. See https://aka.ms/dotnet-JS-interop-threads");
-            }
-#endif
-        }
-
-        private JSSynchronizationContext(Thread targetThread, IntPtr targetTID, WorkItemQueueType queue)
+        internal JSSynchronizationContext(bool isMainThread)
         {
-            TargetThread = targetThread;
-            TargetTID = targetTID;
-            Queue = queue;
+            ProxyContext = new JSProxyContext(isMainThread, this);
+            Queue = Channel.CreateUnbounded<WorkItem>(new UnboundedChannelOptions { SingleWriter = false, SingleReader = true, AllowSynchronousContinuations = true });
             _DataIsAvailable = DataIsAvailable;
+
+            previousSynchronizationContext = Current;
+            SetSynchronizationContext(this);
         }
 
         public override SynchronizationContext CreateCopy()
         {
-            return new JSSynchronizationContext(TargetThread, TargetTID, Queue);
+            // child thread will inherit this JSSynchronizationContext
+            return this;
```

Comment:
this also feels like asking for trouble. under what circumstances does CreateCopy get called? What assumptions are violated if CreateCopy doesn't create a copy?

Reply:
it probably makes difference when you compare them by reference. And probably they should not be the same.